devito/types/tensor.py

```diff
@@ -221,6 +221,8 @@ def div(self, shift=None):
         func = vec_func(self, self)
         for j, d in enumerate(self.space_dimensions):
             comps.append(sum([getattr(self[j, i], 'd%s' % d.name)
+                              (x0=None if shift is None else d + shift[i] * d.spacing
```

Comment:
If we wanna go that way then it needs to be a tuple of tuple for a tensor, not just a tuple.

Reply:
That's only valid for a vector with one component per dimension. A tensorfunction is a `ndimxndim` matrix with one component per dimension pair. It's the same thing as staggerring setup, it is a tuple for Vector and a tuple of tuple for Tensor

Reply:
Nice!! I'm gonna look into that..

Reply:
Please review again. I think this is a good solution. We just need to visit shift values once. One thing to note here is that you can pass (ndim,) or (ndim,ndim) shaped shift (beyond a single float) when dealing with tensors (grad of vectors and div of tensors). A (a,b,c) tuple would be considered as ((a,b,c), (a,b,c), (a,b,c)). Let me know if this would be undesirable.

Reply:
>A (a,b,c) tuple would be considered as ((a,b,c), (a,b,c), (a,b,c)). Let me know if this is undesirable.

That's not a good default IMO. In general, these tenors are symmetric/antisymmetric at least in the staggering so the shift should be too. There is not "easy" way to go from a tuple to a tuple of tuple for tensors and only tuple of tuples should be accepted I think.

Reply:
Would have a single function yes. May be easier and more compact with some  try/catch to merge all cases.